### PR TITLE
CPP: Fix FPs from detecting commented out preprocessor logic

### PR DESCRIPTION
**Pull request by geoffw0**
Fixes the FPs introduced by https://github.com/Semmle/ql/pull/1223.  I'll start a CPP-Differences build to confirm this.

### CODE REVIEW

**Comment by geoffw0** [comment]
CPP-Differences running now, on the changes from both https://github.com/Semmle/ql/pull/1223/ and this PR: https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/551/

(I'm expecting a fair number of true positive changes to remain)

File: cpp/ql/src/Documentation/CommentedOutCode.qll [hunk 1]
```diff
@@ -47,12 +47,27 @@ private predicate looksLikeCode(string line) {
...
 /**
  * The line of a C++-style comment within its file `f`.
  */
 private int lineInFile(CppStyleComment c, File f) {
   f = c.getFile() and
-  result = c.getLocation().getStartLine()
+  result = c.getLocation().getStartLine() and
+
+  // Ignore comments on the same line as a preprocessor directive.
+  not preprocLine(f, result)
```
Comment (jbj) [inline cpp/ql/src/Documentation/CommentedOutCode.qll]: Why make this exception only for C++-style comments? This pattern is used in at least one C-style comment in our own codebase. You can find it with `git grep '#endif.*/\* #'` from Bash.

File: cpp/ql/src/Documentation/CommentedOutCode.qll [hunk 2]
```diff
@@ -47,12 +47,27 @@ private predicate looksLikeCode(string line) {
   )
 }
 
+/**
+ * Holds if there is a preprocessor directive on the line indicated by
+ * `f` and `line`.
+ */
+private predicate preprocLine(File f, int line) {
+  exists(PreprocessorDirective pd, Location l |
```
Comment (jbj) [inline cpp/ql/src/Documentation/CommentedOutCode.qll]: Shouldn't this be limited to `#else` and `#endif`? If not, please add a test with an example.

File: cpp/ql/src/Documentation/CommentedOutCode.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Documentation/CommentedOutCode.qll]: This logic actually covers an edge case.  The main exclusion is in the constructor of `CommentBlock`, to catch comments that begin on the same line as a preprocessor directive, e.g.:
```
#endif /* #ifdef MYMACRO */
```
or indeed:
```
#endif /* #ifdef MYMACRO
more comment here
*/
```

The edge case involves CPP-style comments as the query has logic that merges consecutive CPP-style comments into one.  For example:
```
// first
// second
#endif // third
// fourth
```
is considered a single four-line comment.  The logic here eliminates the line `//third` _before_ merging, so that the above is now considered to be two comments (one consisting of the first two lines, and one of just the fourth).  There's a test case for this, admittedly narrow, behaviour.

Having said all this it looks like I forgot to add a test case for a CPP-style comment.  I'll do that now.

Comment (geoffw0) [reply cpp/ql/src/Documentation/CommentedOutCode.qll]: Done.

File: cpp/ql/src/Documentation/CommentedOutCode.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/Documentation/CommentedOutCode.qll]: Yes I think you're right.  Or `#elif`.  Updated.

**Comment by geoffw0** [comment]
Could you let me know if you want to see an updated differences job (following the restriction to `#elif`, `#else`, `#endif`?  I'm inclined to just leave that one running and rely on the daily job to check any odd cases that are different.

**Comment by geoffw0** [comment]
Well, the differences job has failed, for mysterious reasons.  I've started another one which includes the new changes: https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/552/

**Comment by geoffw0** [comment]
Third attempt: https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/554/

**Comment by geoffw0** [comment]
Diff job above succeeded.  There are lots of new results, but the ones I sampled all look correct now.

**Comment by rdmarsh2** [comment]
I saw one new FN in that dashboard:
chakracore_windows | > | lib | > | Common | > | PlatformAgnostic | > | Debugger.h

**Comment by geoffw0** [comment]
> I saw one new FN in that dashboard:
chakracore_windows | > | lib | > | Common | > | PlatformAgnostic | > | Debugger.h

Well spotted.  I've just done a bit of investigation and concluded that `Debugger.h` actually contains no code, due to the `#ifndef _WIN32`, and thus is excluded by the new special case for header files (such as `config.h` files) that contain no code.  It's slightly weird but we _do_ see comments inside not-taken preprocessor branches even though we don't see the code (due to the way the extractor works), which is why we were able to produce a result there in the first place.

In this case I'd probably prefer if we kept the result, but we've only lost one good result compared to hundred of bad ones from actual `config.h` and other special files, and I can't think of an easy way to discriminate this one.

I've added a new test case showing the case of commented out code inside a not-taken preprocessor branch, because I think this behaviour is not obvious.
